### PR TITLE
Add cryptoCb hook to one-shot CMAC functions

### PR DESCRIPTION
**Pull request by bigbrett**
- Adds cryptoCb hook to one-shot CMAC functions 
- add CMAC coverage to cryptoCb tests

# Testing

Added CMAC cryptoCb to `myCryptoDevCb` to exercise new and existing code paths

### CODE REVIEW

File: wolfcrypt/test/test.c [hunk 1]
```diff
@@ -39968,14 +39968,26 @@ WOLFSSL_TEST_SUBROUTINE wc_test_ret_t cmac_test(void)
 
         XMEMSET(tag, 0, sizeof(tag));
         tagSz = sizeof(tag);
+#if !defined(HAVE_FIPS) || \
```
Comment (billphipps) [inline wolfcrypt/test/test.c:39971]: Hrmmm... Not sure why this test is limited by FIPS.  Shouldn't we be performing the _ex test when cryptocb is defined rather than FIPS?  I haven't reviewed the other chunks of this test code. Currently, FIPS is not compatible with cryptocb so we may need to add a second test that explicitly works the cryptocb version (devid!=INVALID_DEVID) of this code .

File: wolfcrypt/test/test.c [hunk 2]
```diff
@@ -49368,6 +49380,49 @@ static int myCryptoDevCb(int devIdArg, wc_CryptoInfo* info, void* ctx)
         info->hmac.hmac->devId = devIdArg;
     }
 #endif
+#if defined(WOLFSSL_CMAC) && !defined(NO_AES) && defined(WOLFSSL_AES_DIRECT)
```
Comment (billphipps) [inline wolfcrypt/test/test.c:49383]: Why is AES_DIRECT required here?

File: wolfcrypt/test/test.c [hunk 3]
```diff
@@ -49368,6 +49380,49 @@ static int myCryptoDevCb(int devIdArg, wc_CryptoInfo* info, void* ctx)
...
+
+        if (info->cmac.in != NULL && info->cmac.out != NULL) {
+            ret = wc_AesCmacGenerate(info->cmac.out,
+                                     info->cmac.outSz,
+                                     info->cmac.in,
+                                     info->cmac.inSz,
+                                     info->cmac.key,
+                                     info->cmac.keySz);
+        }
+        else if (info->cmac.key != NULL) {
```
Comment (billphipps) [inline wolfcrypt/test/test.c:49400]: These cases are not mutually exclusive but rather sequential. Change the non-oneshot logic to:
ret = 0;
if ((ret == 0) && (key !=NULL)) ret = InitCmac(...);
if ((ret == 0) && (in !=NULL)) ret = UpdateCmac(...);
if ((ret == 0) && (out !=NULL)) ret = FinalCmac(...);

**Review by billphipps — COMMENTED** [review]
Minor nits and some ideas.  Feel free to resolve at your discretion.

File: wolfcrypt/test/test.c (hunk 3 above)
Comment (bigbrett) [reply wolfcrypt/test/test.c:49400]: hmm I don't understand why they should be sequential rather than mutually exclusive. Isn't the point for there to be a 1-to-1 mapping between the WC API called and the case invoked here (e.g. the cryptoCb detecting whether it is being invoked from init/update/final)?  The user can only call one of  init, update, or final, and each of those functions passes a mutually exclusive set of args to the cryptocb (for example, there is no way for `wc_InitCmac` to pass a non-null `in` or `out` arg to the cryptoCb, and no way for `wc_CmacUpdate` to pass a non-null `out` arg to the cryptoCb) so why would we want to support "falling through" to these cases inside the callback?

File: wolfcrypt/test/test.c (hunk 1 above)
Comment (bigbrett) [reply wolfcrypt/test/test.c:39971]: So I just followed the existing pattern of how `wc_InitCmac` vs `wc_initCmac_ex` was called earlier in this test. And to be clear, the `_ex` is called only when FIPS is NOT on, my guess being because FIPS isn't compat with cryptoCb. But I agree, you would think it would be gated based on whether cryptoCb is on, not FIPS, leaving the incompatibility of the two as a compile time detection error.

Elsewhere in `test.c` the FIPS check isn't even present and it just calls the `_ex` versions of crypto functions regardless, relying on the fact that internal cryptoCb invocations will be gated out and `_ex` will behave like normal if cryptoCb support isn't turned on. 

We should probably just remove the FIPS check and call the _ex function regardless, as cryptoCb code will not be compiled if support for it isn't turned on.

> we may need to add a second test that explicitly works the cryptocb version (devid!=INVALID_DEVID) of this code.

This should already be happening. When cryptoCb support is turned on, the `cryptocb_test()` function invokes this function (`cmac_test()`), but with a valid cryptoCb registered. I verified the cryptoCb code paths were being hit.

File: wolfcrypt/src/cmac.c [hunk 4]
```diff
@@ -326,14 +328,40 @@ int wc_AesCmacGenerate(byte* out, word32* outSz,
...
+    /* Set devId regardless of value (invalid or not) */
+    cmac->devId = devId;
+    #ifndef WOLF_CRYPTO_CB_FIND
+    if (devId != INVALID_DEVID)
+    #endif
+    {
+        cmac->devCtx = NULL;
+
+        ret = wc_CryptoCb_Cmac(cmac, key, keySz, in, inSz, out, outSz,
+                WC_CMAC_AES, NULL);
```
[the remark this answers is not in the document]
Comment (bigbrett) [reply wolfcrypt/src/cmac.c:350]: Since [existing code in wc_InitCmac](https://github.com/wolfSSL/wolfssl/blob/b7b20ededda4cea208fb7745629904fda64c7524/wolfcrypt/src/cmac.c#L107) checks against WC_CMAC_AES for validity, I assumed it was supposed to hold a `CmacType` enum. And if we ever do CMAC with a different block cypher in the future, then you would multiplex between implementations based on that field. But our CMAC impl is hard-coded for AES at the moment so it is kind of pointless.

>Looking at the other callback invocations, it seems like it is set to 0 except during the init. 

I imagine this is just because only the init functions even take a type argument, so go unused during update/final

I can pass it 0 for this callback to maintain consistency

File: wolfcrypt/src/cmac.c [hunk 5]
```diff
@@ -326,14 +328,40 @@ int wc_AesCmacGenerate(byte* out, word32* outSz,
...
+        }
+        /* fall-through when unavailable */
+    }
+#else
+    (void)devId;
+#endif
+
     ret = wc_InitCmac(cmac, key, keySz, WC_CMAC_AES, NULL);
     if (ret == 0) {
         ret = wc_CmacUpdate(cmac, in, inSz);
```
[the remark this answers is not in the document]
Comment (bigbrett) [reply wolfcrypt/src/cmac.c:367]: good catch. Looking at it, I think we need to go further and actually use the new wc_CmacFree in this case, which ensures the CMAC AES context is freed then does the zero. Changed.

File: wolfcrypt/src/cmac.c [hunk 6]
```diff
@@ -369,7 +405,7 @@ int wc_AesCmacVerify(const byte* check, word32 checkSz,
     }
 
     XMEMSET(a, 0, aSz);
-    ret = wc_AesCmacGenerate(a, &aSz, in, inSz, key, keySz);
+    ret = wc_AesCmacGenerate_ex(a, &aSz, in, inSz, key, keySz, devId);
     compareRet = ConstantCompare(check, a, (int)min(checkSz, aSz));
```
[the remark this answers is not in the document]
Comment (bigbrett) [reply wolfcrypt/src/cmac.c:409]: Fixed ret check. 

Regarding the cryptoCb, that is a great idea, but we would need to add additional info to the cryptoCb so that it could determine the difference between generate and verify. If the `type` field wasn't intended to be used to indicate the block cipher type, then we could use the field for that (as well as init/update/final). Otherwise we could add a new "operation" field to indicate this.

File: wolfcrypt/test/test.c (hunk 1 above)
Comment (bigbrett) [reply wolfcrypt/test/test.c:39971]: Figured it out - the call is gated by that particular version of FIPS because at that point in time, the API did not exist. So when our Jenkins FIPS tests run (and replace cmac.{c,h} with their FIPS implementations from the FIPS repo) then this wolfCrypt test will be missing that symbol.

File: wolfcrypt/src/cmac.c [hunk 7]
```diff
@@ -354,28 +408,87 @@ int wc_AesCmacGenerate(byte* out, word32* outSz,
...
+                        int type, void* heap, int devId)
 {
     int ret;
     byte a[AES_BLOCK_SIZE];
     word32 aSz = sizeof(a);
     int compareRet;
 
-    if (check == NULL || checkSz == 0 || (in == NULL && inSz != 0) ||
-        key == NULL || keySz == 0) {
+    if (check == NULL || checkSz == 0 || (in == NULL && inSz != 0)) {
```
Comment (billphipps) [inline wolfcrypt/src/cmac.c:422]: Add parameter check (cmac == NULL)

File: wolfcrypt/src/cmac.c [hunk 8]
```diff
@@ -305,6 +305,58 @@ int wc_CmacFinal(Cmac* cmac, byte* out, word32* outSz) {
     return ret;
 }
 
+
+int wc_AesCmacGenerate_ex(Cmac* cmac,
+                          byte* out, word32* outSz,
+                          const byte* in, word32 inSz,
+                          const byte* key, word32 keySz,
+                          int type, void* heap, int devId)
```
Comment (billphipps) [inline wolfcrypt/src/cmac.c:313]: Just thought of this.  The function is AEScmacgenerate, so why do we accept the "type" parameter?  Maybe we should drop it and simply force it to WC_CMAC_AES in the init call below.  Same with the verify.  Otherwise, we really should check that it is the value we expect OR we expose generic version of cmacgenerate that is invoked by AEScmacgenerate.

File: wolfcrypt/test/test.c [hunk 9]
```diff
@@ -39968,16 +39968,42 @@ WOLFSSL_TEST_SUBROUTINE wc_test_ret_t cmac_test(void)
...
+                             tc->k, tc->kSz, WC_CMAC_AES, NULL, devId);
+#else
         ret = wc_AesCmacVerify(tc->t, tc->tSz, tc->m, tc->mSz,
                              tc->k, tc->kSz);
+#endif
         if (ret != 0)
             ERROR_OUT(WC_TEST_RET_ENC_EC(ret), out);
+
+#if !defined(HAVE_FIPS) || FIPS_VERSION_GE(5, 3)
+        /* Test that keyless generate with init is the same */
```
Comment (billphipps) [inline wolfcrypt/test/test.c:39993]:
```suggestion
        /* Test that pre-keyed generate with init is the same */
```

File: wolfcrypt/test/test.c [hunk 10]
```diff
@@ -39968,16 +39968,42 @@ WOLFSSL_TEST_SUBROUTINE wc_test_ret_t cmac_test(void)
...
+        XMEMSET(tag, 0, sizeof(tag));
+        tagSz = sizeof(tag);
+        ret = wc_InitCmac_ex(cmac, tc->k, tc->kSz, tc->type, NULL, HEAP_HINT, devId);
+        if (ret != 0) {
+            ERROR_OUT(WC_TEST_RET_ENC_EC(ret), out);
+        }
+        ret = wc_AesCmacGenerate_ex(cmac, tag, &tagSz, tc->m, tc->mSz,
+                NULL, tc->kSz, WC_CMAC_AES, NULL, devId);
```
Comment (billphipps) [inline wolfcrypt/test/test.c:40001]:
```suggestion
                NULL, tc->kSz, WC_CMAC_AES, HEAP_HINT, devId);
```

File: wolfcrypt/test/test.c [hunk 11]
```diff
@@ -49368,6 +49394,49 @@ static int myCryptoDevCb(int devIdArg, wc_CryptoInfo* info, void* ctx)
...
+#if defined(WOLFSSL_CMAC) && !defined(NO_AES)
+    else if (info->algo_type == WC_ALGO_TYPE_CMAC) {
+        if (info->cmac.cmac == NULL) {
+            return NOT_COMPILED_IN;
+        }
+
+        /* set devId to invalid so software is used */
+        info->cmac.cmac->devId = INVALID_DEVID;
+
+        /* Handle one-shot cases */
```
Comment (billphipps) [inline wolfcrypt/test/test.c:49406]: Should we add the case of key==null && in!=NULL &out!=null to directly invoke generate_ex?  Not really needed, but could show a decent model for a real cryptocb handler

File: wolfssl/wolfcrypt/cmac.h [hunk 12]
```diff
@@ -111,11 +111,27 @@ WOLFSSL_API
...
+int wc_AesCmacVerify_ex(Cmac* cmac,
+                        const byte* check, word32 checkSz,
+                        const byte* in, word32 inSz,
+                        const byte* key, word32 keySz,
+                        int type,
+                        void* heap,
+                        int devId);
 
 WOLFSSL_LOCAL
 void ShiftAndXorRb(byte* out, byte* in);
```
Comment (billphipps) [inline wolfssl/wolfcrypt/cmac.h:137]: Hrmm... Why is this here?  if this is static, then the declaration should be in the source file, not in this public header.  No change necessary.

**Review by billphipps — CHANGES_REQUESTED** [review]
A few missed parameter checks and a couple of questions.  Feel free to resolve at your discretion and I'll quickly approve.

File: wolfcrypt/src/cmac.c [hunk 13]
```diff
@@ -305,6 +305,58 @@ int wc_CmacFinal(Cmac* cmac, byte* out, word32* outSz) {
...
+    ret = 0;
+
+    /* Init step is optional */
+    if (key != NULL) {
+        ret = wc_InitCmac_ex(cmac, key, keySz, type, NULL, heap, devId);
+    }
+    if (ret == 0) {
+        ret = wc_CmacUpdate(cmac, in, inSz);
+        /* Ensure we are freed and zeroed if not calling wc_CmacFinal */
+        if (ret != 0) {
```
[the remark this answers is not in the document]
Comment (bigbrett) [reply wolfcrypt/src/cmac.c:348]: So the original generate function wasn't suitable for async use, since it allocated/freed on entry/exit, and even though this new generate_ex version takes a context, I still don't think its async-capable as it would still be calling "init" on every invocation. 

I added this line because you asked for the context to be zeroed in an earlier comment and I figured if we were completely clearing the context (e.g. zeroing the aes struct) then we probably wanted to free the AES substruct as well. But I wasn't thinking about this function being used in an async context, given that it is a "one shot". 

Also, the CMAC context doesn't hold any sort of async state, only the underlying AES module. So there is no way to "poll" the underlying state of the AES operation and resume using this function.

Given this, do you still want me to clear it on failure but not free? Or if you think it should be not async capable, then is the free okay?

File: wolfcrypt/src/cmac.c (hunk 8 above)
Comment (bigbrett) [reply wolfcrypt/src/cmac.c:313]: I just followed the signature of the other _ex invocations, but I agree about the hard-coded AES dependencies. Removed.

File: wolfcrypt/test/test.c (hunk 11 above)
Comment (bigbrett) [reply wolfcrypt/test/test.c:49406]: This is true, though I don't like the idea of adding additional stuff into the test cryptoCb handler that isn't actually being tested (as these are just unit tests with no ability to use the key loaded out-of-band). I think it would be best to have a reference implementation of a cryptoCb in documentation that could handle all the cases. 

We also now have more "operations" than our input combinations can support (if we wanted the verify operation to be triggerable as a callback), so we should probably add a bitfield "comand" argument vs deducing operations based on input parameters.

I think this should be saved for a subsequent PR

File: wolfcrypt/src/cmac.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (billphipps) [reply wolfcrypt/src/cmac.c:360]: I think we need to thin about if/how cryptocb could support async responses.  Keep the code the same for now.

**Review by billphipps — APPROVED** [review]

**Comment by bigbrett** [comment]
Jenkins retest this please

**Comment by bandi13** [comment]
retest this please

**Review by dgarske — APPROVED** [review]
